### ISSUE TITLE
QueryDSL query object initialization can lead to deadlock if default variables are used

### ISSUE THREAD

**Issue by vanti1980** [querydsl/querydsl#2334]
There are 2 entities: User and Mandant.
User has a field referring to Mandant and Mandant has a field referring to User.

For both entities it has not been turned off to generate default variables therefore the following query objects are generated:

[querydsl-query-object-initialization-bug.zip](https://github.com/querydsl/querydsl/files/2167115/querydsl-query-object-initialization-bug.zip)

Let's assume that two queries are started in parallel (in different threads):

    public List<Mandant> getAllMandants()
    {
	QMandant qMandant = new QMandant("m");
        //...
    }

    public List<User> getAllUsers()
    {
	QUser qUser = new QUser ("u");
        //...
    }

In a very-very-very unlikely situation the following can occur:
"sched_clustered_Worker-18" #66 prio=4 os_prio=0 tid=0x00007f84301e0000 nid=0x3c50 in Object.wait() [0x00007f83ffcfa000]
   java.lang.Thread.State: RUNNABLE
	at com.example.domain.generated.QMandant.\<init\>(QMandant.java:131)
	at com.example.domain.generated.QMandant.\<init\>(QMandant.java:120)
	at com.example.domain.generated.QMandant.\<init\>(QMandant.java:130)
	at com.example.domain.generated.QMandant.\<init\>(QMandant.java:108)
	at com.example.domain.generated.QMandant.\<clinit\>(QMandant.java:23)
	at com.example.dao.jpa.MandantDao.getAllMandants(MandantDao.java:30)

"sched_clustered_Worker-17" #65 prio=4 os_prio=0 tid=0x00007f84301df000 nid=0x3c4f in Object.wait() [0x00007f83ffdfa000]
   java.lang.Thread.State: RUNNABLE
	at com.example.domain.generated.QUser.\<init\>(QUser.java:71)
	at com.example.domain.generated.QUser.\<init\>(QUser.java:54)
	at com.example.domain.generated.QUser.\<clinit\>(QUser.java:23) 
	at com.example.dao.jpa.UserDao.getAllUsers(UserDao.java:30)

The reason for that:
- class initialization (which is guaranteed to be sequential in JLS) of QUser cannot be completed because it wants to create a QMandant (before that QMandant.class has to be initialized)
- class initialization of QMandant cannot be completed because it wants to create a QUser (before that QUser.class has to be initialized)

The problem is actually the default variable created which is immediately initialized with a fresh instance. 

Workaround: if you turn it off then class initializer doesn't want to create an instance of the corresponding class and thus does not want to run the class initializer for that. JVM-experts please check if I'm right about that!

**Comment by Shredder121** [comment]
Another workaround is to ensure one place at least once references the default variables.
Then one thread will initialize the classes, after which it's all good.
There isn't really a way around this, unless you'd be fine with accessors for the properties instead of fields.

**Comment by mgraciano** [comment]
Actually this issue can happen just using a j.u.Date JPA mapping with QueryDSL 3 (Tested with QueryDSL 3.3.2). I suspect the problem is in class DateTimeExpression references to DateTimeOperation, which is subclass of DateTimeExpression, causing the system to hangs.
We were able to reproduce it in environments with some concurrent requests. As a temporary workaround, we initialized the classes manually in a single thread fashion. This particular mapping was wrongly using j.u.Date, so the final fix will be changing it to j.t.LocalDateTime. I was able to find a JDK issue, https://bugs.openjdk.java.net/browse/JDK-8037567, which describes a similar scenario.

```java
@Startup
@Singleton
public class StartupBean {

    private static final Logger LOGGER = Logger.getLogger(StartupBean.class);

    @PostConstruct
    public void init() {
        LOGGER.warn("Initializing...");
        DateTimeOperation.currentDate();
        DateTimeExpression.currentDate();
        LOGGER.warn("Initialized");
    }

}
```
